enhance: add disk file writer with Direct IO support

sparknack opened the pull request:
issue: #43040 

This patch introduces a disk file writer that supports Direct IO.

Currently, it is exclusively utilized during the QueryNode load process.

Below is its parameters:

1. `common.diskWriteMode`
This parameter controls the write mode of the local disk, which is used to write temporary data downloaded from remote storage.
Currently, only QueryNode uses 'common.diskWrite*' parameters. Support for other components will be added in the future.
The options include 'direct' and 'buffered'. The default value is 'buffered'.

2. `common.diskWriteBufferSizeKb`
Disk write buffer size in KB, only used when disk write mode is 'direct', default is 64KB.
Current valid range is [4, 65536]. If the value is not aligned to 4KB, it will be rounded up to the nearest multiple of 4KB.

3. `common.diskWriteNumThreads`
This parameter controls the number of writer threads used for disk write operations. The valid range is [0, hardware_concurrency].
It is designed to limit the maximum concurrency of disk write operations to reduce the impact on disk read performance.
For example, if you want to limit the maximum concurrency of disk write operations to 1, you can set this parameter to 1.
The default value is 0, which means the caller will perform write operations directly without using an additional writer thread pool.
In this case, the maximum concurrency of disk write operations is determined by the caller's thread pool size.

Both parameters can be updated during runtime.

sparknack committed:
remove unused TEST_skip_index_for_retrieve param in CreateSealedSegment

Signed-off-by: Shawn Wang <shawn.wang@zilliz.com>

sparknack committed:
remote unused header files

Signed-off-by: Shawn Wang <shawn.wang@zilliz.com>

sparknack committed:
clean some unused code

Signed-off-by: Shawn Wang <shawn.wang@zilliz.com>

sparknack committed:
enhance: add disk file writer with Direct IO support

Signed-off-by: Shawn Wang <shawn.wang@zilliz.com>

sparknack commented:
rerun cpp-unit-test

sparknack commented:
rerun cpp-unit-test

sparknack commented:
rerun cpp-unit-test

sparknack commented:
/hold

liliu-z commented on internal/core/src/common/ChunkWriter.h:
```diff
@@ -73,8 +76,8 @@ class ChunkWriter final : public ChunkWriterBase {
     ChunkWriter(int dim, bool nullable) : ChunkWriterBase(nullable), dim_(dim) {
     }
 
```
Seems no place is using this offset, any ctx why we have this before. Maybe @sunby has some ideas

liliu-z commented on internal/core/src/storage/FileWriter.h:
```diff
@@ -0,0 +1,134 @@
+    size_t offset_{0};
+};
+
+class FileWriterConfig {
```
It pretty weird to me to have a singleton here. Why we just use a global parameters read from yml file?

liliu-z commented on internal/core/src/storage/FileWriter.h:
```diff
@@ -0,0 +1,134 @@
+
+namespace milvus::storage {
+
+class FileWriter {
```
Looks like this is not thread safe, plz leave a comment for this. Also did you get a chance to ensure we didn't use it in a concurrent surroundings.

liliu-z reviewed (commented):
Nice work

sparknack committed:
update of disk file writer implementation

1. add an extra folly thread pool to handle data writing.
2. add new unittests.

Signed-off-by: Shawn Wang <shawn.wang@zilliz.com>

sparknack committed:
use a new implementation of disk write worker pool

Signed-off-by: Shawn Wang <shawn.wang@zilliz.com>

sparknack committed:
some new update

Signed-off-by: Shawn Wang <shawn.wang@zilliz.com>

zhengbuqian commented on pkg/util/paramtable/component_param.go:
```diff
@@ -649,6 +653,38 @@ This configuration is only used by querynode and indexnode, it selects CPU instr
 
+	p.DiskWriteMode = ParamItem{
+		Key:          "common.diskWriteMode",
+		Version:      "2.5.14",
```
2.6.0

zhengbuqian commented on internal/querynodev2/server.go:
```diff
@@ -201,13 +201,29 @@ func ResizeHighPriorityPool(evt *config.Event) {
+		mode := pt.CommonCfg.DiskWriteMode.GetValue()
+		bufferSize := pt.CommonCfg.DiskWriteBufferSizeKb.GetAsUint64()
+		nrThreads := pt.CommonCfg.DiskWriteNrThreads.GetAsInt()
+		C.InitFileWriterConfig(C.CString(mode), C.uint64_t(bufferSize), C.int(nrThreads))
```
handle error

zhengbuqian commented on internal/querynodev2/server.go:
```diff
@@ -241,6 +257,13 @@ func (node *QueryNode) InitSegcore() error {
 	C.InitLowPriorityThreadCoreCoefficient(cLowPriorityThreadCoreCoefficient)
+
+	cDiskWriteMode := C.CString(paramtable.Get().CommonCfg.DiskWriteMode.GetValue())
+	cDiskWriteBufferSizeKb := C.uint64_t(paramtable.Get().CommonCfg.DiskWriteBufferSizeKb.GetAsUint64())
```
what if the provided size is not valid/aligned?

zhengbuqian commented on internal/core/src/storage/FileWriter.cpp:
```diff
@@ -0,0 +1,359 @@
+                      e.what());
+        }
+    } else {
+        if (use_direct_io_) {
```
we can simply call task() here?

zhengbuqian commented on internal/core/src/common/ChunkWriter.h:
```diff
@@ -22,14 +23,17 @@
 
-    ChunkWriterBase(File& file, size_t offset, bool nullable)
-        : file_(&file), file_offset_(offset), nullable_(nullable) {
+    ChunkWriterBase(std::string file_path, bool nullable)
```
how about std::string&& file_path

zhengbuqian reviewed (commented):
thank you!

sparknack committed:
refactor and fix

Signed-off-by: Shawn Wang <shawn.wang@zilliz.com>

sparknack commented:
/unhold

liliu-z reviewed (commented):
/lgtm
